### PR TITLE
Convert Field.java to Kotlin

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: wire-codegen-sample/src/main/java/com/squareup/wire/sample/ServiceGenerator.java [hunk 1]
```diff
@@ -45,19 +45,19 @@ public TypeSpec api(Service service) {
     }
 
     for (Rpc rpc : service.rpcs()) {
-      ProtoType requestType = rpc.requestType();
+      ProtoType requestType = rpc.getRequestType();
```
Comment (swankjesse) [inline wire-codegen-sample/src/main/java/com/squareup/wire/sample/ServiceGenerator.java]: A bunch of these accessors should disappear in follow-up when I convert these types to Kotlin.

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/FieldExtensions.kt [hunk 2]
```diff
@@ -23,10 +23,10 @@ import com.squareup.wire.schema.Field
 import com.squareup.wire.schema.ProtoType
 
 internal val Field.isMap: Boolean
-  get() = type().isMap
+  get() = type!!.isMap
```
Comment (swankjesse) [inline wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/FieldExtensions.kt]: ... might change this to be lateinit later. It never escapes the Linker without being non-null.

File: wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+ * fully-qualified name using the protocol buffer package.
+ */
+class ProtoType {
+  val isScalar: Boolean
+
+  private val string: String
+
+  val isMap: Boolean
+
+  /** The type of the map's keys. Only present when [.isMap] is true.  */
```
Comment (Egorand) [inline wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt]: nit: `[isMap]`

File: wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  val keyType: ProtoType?
+
+  /** The type of the map's values. Only present when [.isMap] is true.  */
```
Comment (Egorand) [inline wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt]: nit: `[isMap]`

File: wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  val valueType: ProtoType?
+
+  val simpleName: String
+    get() {
```
Comment (Egorand) [inline wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt]: Why not compute once and assign?

File: wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+      val dot = string.lastIndexOf('.')
+      return string.substring(dot + 1)
+    }
+
+  /** Returns the enclosing type, or null if this type is not nested in another type.  */
+  val enclosingTypeOrPackage: String?
+    get() {
```
Comment (Egorand) [inline wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt]: Same

File: wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+      val dot = string.lastIndexOf('.')
+      return if (dot == -1) null else string.substring(0, dot)
+    }
+
+  /** Creates a scalar or message type.  */
+  private constructor(isScalar: Boolean, string: String) {
+    checkNotNull(string, "string == null")
```
Comment (Egorand) [inline wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt]: Null check is probably extraneous.

File: wire-schema/src/main/java/com/squareup/wire/schema/Field.kt [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+  val isRequired: Boolean
+    get() = label == Label.REQUIRED
+
+  /**
+   * Returns this field's name, prefixed with its package name. Uniquely identifies extension
+   * fields, such as in options.
+   */
+  val qualifiedName: String
+    get() {
+      return when {
```
Comment (Egorand) [inline wire-schema/src/main/java/com/squareup/wire/schema/Field.kt]: Compute once and assign? Same for fields below.

File: wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt (hunk 3 above)
Comment (swankjesse) [reply wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt]: Fixed!

File: wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt (hunk 4 above)
Comment (swankjesse) [reply wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt]: Fixed!

File: wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt (hunk 5 above)
Comment (swankjesse) [reply wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt]: To save allocations. This isn’t usually needed.

File: wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt (hunk 6 above)
Comment (swankjesse) [reply wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt]: I’d want to make that kind of change independent of the language conversion.

File: wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt (hunk 7 above)
Comment (swankjesse) [reply wire-schema/src/main/java/com/squareup/wire/schema/ProtoType.kt]: Fixed. Good catch.

File: wire-schema/src/main/java/com/squareup/wire/schema/Field.kt (hunk 8 above)
Comment (swankjesse) [reply wire-schema/src/main/java/com/squareup/wire/schema/Field.kt]: I think it reads better as computed-on-demand state. Lets discuss offline; I’m don’t hold strong feelings here.

**Pull request merged**